replica/distributed_loader.cc

```diff
@@ -434,28 +451,24 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
         };
         process_sstable_dir(directory, flags).get();
 
-        auto generation = highest_generation_seen(directory).get0();
-        auto shard_generation_base = generation.value_or(replica::table::make_new_generation()).value() / smp::count + 1;
-
-        // We still want to do our best to keep the generation numbers shard-friendly.
-        // Each destination shard will manage its own generation counter.
-        std::vector<std::atomic<sstables::generation_type::int_t>> shard_gen(smp::count);
-        for (shard_id s = 0; s < smp::count; ++s) {
-            shard_gen[s].store(shard_generation_base * smp::count + s, std::memory_order_relaxed);
-        }
-
-        // we need generation calculated by instance of cf at requested shard
-        auto new_generation_for_shard = [&] (shard_id shard) {
-            auto gen = shard_gen[shard].fetch_add(smp::count, std::memory_order_relaxed);
-            return sstables::generation_type(gen);
-        };
-
+        std::vector<std::unique_ptr<sstable_generation_generator>> shard_gen;
+        auto highest_generation = highest_generation_seen(directory).get0().value_or(
+            replica::table::make_new_generation());
+        std::generate_n(std::back_inserter(shard_gen),
+                        smp::count,
+                        [highest_generation] {
+                            uint64_t base = (highest_generation.value() / smp::count + 1) * smp::count;
```

Comment:
it doesn't seem like you need to `+ 1` here since `sstable_generation_generator::operator()` adds smp::count to the last generation

Reply:
i don't think so. what `sstable_generation_generator::operator()` returns is practically
```c++
_last_generation.fetch_add(smp::count, std::memory_order_relaxed) + this_shard_id()
```

and `fetch_add()` returns the original value before being incremented by `smp::count`. and before my change, the integer based generation id starts with `smp::count` , assuming the highest_generation is 0. as we always add 1 to the base before multiply it by `smp::count`.

Reply:
yup, you're right. sorry for the noise